### PR TITLE
Fix behavior of iterate on indices <= 0

### PR DESCRIPTION
**Pull request by cjprybol**
## Types of changes

This PR implements the following changes:
_(Please tick any or all of the following that are applicable)_

* [ ] :sparkles: New feature (A non-breaking change which adds functionality).
* [x] :bug: Bug fix (A non-breaking change, which fixes an issue).
* [ ] :boom: Breaking change (fix or feature that would cause existing functionality to change).

## :clipboard: Additional detail

This is how iterate behaves on strings
```julia
julia> x = "ATCG"
"ATCG"

julia> iterate(x)
('A', 2)

julia> iterate(x, -1)

julia> iterate(x, 0)

julia> iterate(x, 1)
('A', 2)

julia> iterate(x, 2)
('T', 3)

julia> iterate(x, 4)
('G', 5)

julia> iterate(x, 5)
```

iterate currently behaves as expected when indexing >= length(seq), but will crash when using any index less than or equal to the initial position (1)
```julia
julia> x = BioSequences.randdnaseq(3)
3nt DNA Sequence:
TGG

julia> iterate(x, 4)

julia> iterate(x, 0)

signal (11): Segmentation fault
in expression starting at REPL[5]:1
>> at ./int.jl:456 [inlined]
extract_encoded_element at /home/jovyan/workspace/BioSequences.jl/src/bit-manipulation/bitindex.jl:74 [inlined]
extract_encoded_element at /home/jovyan/workspace/BioSequences.jl/src/longsequences/indexing.jl:18 [inlined]
getindex at /home/jovyan/workspace/BioSequences.jl/src/biosequence/indexing.jl:39 [inlined]
iterate at /home/jovyan/workspace/BioSequences.jl/src/biosequence/indexing.jl:11
unknown function (ip: 0x7f5e879a9631)
```

## :ballot_box_with_check: Checklist

- [ ] :art: The changes implemented is consistent with the [julia style guide](https://docs.julialang.org/en/v1/manual/style-guide/).
- [ ] :blue_book: I have updated and added relevant docstrings, in a manner consistent with the [documentation styleguide](https://docs.julialang.org/en/v1/manual/documentation/).
- [ ] :blue_book: I have added or updated relevant user and developer manuals/documentation in `docs/src/`.
- [x] :ok: There are unit tests that cover the code changes I have made.
- [x] :ok: The unit tests cover my code changes AND they pass.
- [ ] :pencil: I have added an entry to the `[UNRELEASED]` section of the manually curated `CHANGELOG.md` file for this repository.
- [ ] :ok: All changes should be compatible with the latest stable version of Julia.
- [ ] :thought_balloon: I have commented liberally for any complex pieces of internal code.

## Notes

Found this while trying to fix https://github.com/BioJulia/Kmers.jl/issues/33 - I think I've got that issue fixed, but this (unrelated issue) was the next failing test that I hit in the Kmers.jl tests

### CODE REVIEW

**Comment by jakobnissen** [comment]
I'm not crazy about this change TBH. The reason is that the second argument to iterate (the state) is generally regarded as an internal object. Using iterate with your own state is, in general in Julia, undefined behaviour. For example:
```julia
julia> s = Set([1,2])
Set{Int64} with 2 elements:
  2
  1

julia> iterate(s, -10000000)
(1392277128300, -9999998)
```
So there is really no issue here to fix. And adding one more branch in a function as low-level as iterate does have some performance implications.

You _could_ change the iteration to have a similar implementation as that of `Array`, where it reinterprets to `UInt`, then exploits the fact that negative unsigned overflow in order to skip the extra check.

**Comment by cjprybol** [comment]
Thanks for the feedback!

I was thinking of BioSequences being most similar to strings and/or arrays, and so those were the 2 types that I tried to mimic the behavior of
```julia
# strings are in the initial post above, arrays here
julia> x = [1]
1-element Vector{Int64}:
 1

julia> iterate(x, 0)

julia> iterate(x, 1)
(1, 2)

julia> iterate(x, 2)

```

I didn't try any other collections, but I see your point though that there isn't a universal standard for how to handle the scenario of iterating at state <= 0, and returning undefined behavior is just as valid as handling the bounds checking.

I like the idea of mimicking the reference implementation in Base as you mentioned:
```julia
julia> @enter iterate(x, 0)
[ Info: tracking Base
In iterate(A, i) at array.jl:781
>781  iterate(A::Array, i=1) = (@_inline_meta; (i % UInt) - 1 < length(A) ? (@inbounds A[i], i + 1) : nothing)
``` 

But just to confirm, when you say
> So there is really no issue here to fix.

Do you agree that we should fix the segfault crash?

If so, it sounds like you want to make sure the updates preserve performance rather than comply 100% with the behavior of arrays and strings. In which case trying the implementations in base, which have likely been optimized and benchmarked, is the best place to start

**Comment by jakobnissen** [comment]
What I mean is that supplying your own state to iterate is undefined behaviour so it doesn't matter if it segfaults. However, since the implementation of iterate for Array manages to be safe with zero cost, we might as well use that, since there is no downside

File: src/biosequence/indexing.jl [hunk 1]
```diff
@@ -8,7 +8,7 @@
 ### License is MIT: https://github.com/BioJulia/BioSequences.jl/blob/master/LICENSE.md
 
 @inline function Base.iterate(seq::BioSequence, i::Int = firstindex(seq))
-    i > lastindex(seq) ? nothing : @inbounds seq[i], i + 1
+    (i % UInt) - 1 < lastindex(seq) ? (@inbounds seq[i], i + 1) : nothing
```
Comment (jakobnissen) [inline src/biosequence/indexing.jl:11]:
```suggestion
    (i % UInt) - 1 < (lastindex(seq) % UInt) ? (@inbounds seq[i], i + 1) : nothing
```

Comment (jakobnissen) [reply src/biosequence/indexing.jl:11]: The UInt/Int comparison adds an extra branch. Don't know how arrays avoids this, but we can elide it like this

**Comment by jakobnissen** [comment]
Thanks for your contribution!

**Pull request merged**